File: Solutions/2025/F#/Libraries/Day1/Library.fs
module Day1

open System.IO
open System.Text.RegularExpressions

type Direction = 
    | Left
    | Right

type Rotation = {
    Direction: Direction
    Amount: int
}

let ParseInput filepath = 
    let parseLine line = 
        let regexMatch = Regex(@"([L|R])(\d+)").Match(line)
        let direction = 
            if regexMatch.Groups[1].Value.ToUpper() = "R" then 
                Right
            else 
                Left
        let amount = int regexMatch.Groups[2].Value
        {Rotation.Direction = direction; Amount = amount}

    File.ReadLines(filepath)
    |> Seq.map parseLine
    |> Seq.toList

let Part1 input = 
    let rotationsFolder acc nextRotation = 
        let rotationAmount = 
            match nextRotation.Direction with 
            | Left -> -1 * nextRotation.Amount
            | Right -> nextRotation.Amount

        (((acc + rotationAmount) % 100) + 100) % 100

    List.scan rotationsFolder 50 input
    |> List.sumBy(fun pos -> if pos = 0 then 1 else 0)

let Part2 input = 
    
    let rec performRotations remainingRotations numZeros currPos = 
        if List.isEmpty remainingRotations then 
            numZeros
        else
            let nextRotation = List.head remainingRotations
            if nextRotation.Amount >= 100 then 
               let remainingFront = {Rotation.Amount = nextRotation.Amount - 100; Direction = nextRotation.Direction}
               let nextRemaining = remainingFront :: (List.tail remainingRotations)
               performRotations nextRemaining (numZeros + 1) currPos
            else 
                let nextPosRaw = 
                    match nextRotation.Direction with 
                    | Left -> currPos - nextRotation.Amount
                    | Right -> currPos + nextRotation.Amount
                if nextPosRaw = 0 then 
                    performRotations (List.tail remainingRotations) (numZeros + 1) nextPosRaw
                else if nextPosRaw < 0 && not (currPos = 0) then
                    performRotations (List.tail remainingRotations) (numZeros + 1) (nextPosRaw + 100)
                else if nextPosRaw < 0 && (currPos = 0) then 
                    performRotations (List.tail remainingRotations) (numZeros) (nextPosRaw + 100)
                else if nextPosRaw >= 100 then 
                    performRotations (List.tail remainingRotations) (numZeros + 1) (nextPosRaw - 100)
                else
                    performRotations (List.tail remainingRotations) numZeros nextPosRaw

    performRotations input 0 50
    


    

File: Solutions/2025/C#/2025/Libraries/Day1/Day1/Library.cs
using System.Security.AccessControl;
using System.Text.RegularExpressions;
using static Direction;

public enum Direction
{
    Left,
    Right
}

namespace Day1
{
    
    public static class Day1
    {
        public record Rotation(Direction direction, int amount);

        public static Rotation ParseLine(string line)
        {
            var regexString = @"([L|R])(\d+)";
            var match = Regex.Match(line, regexString);

            var amount = int.Parse(match.Groups[2].Value);

            if (match.Groups[1].Value == "R")
            {
                return new Rotation(Right, amount);
            }
            else
            {
                return new Rotation(Left, amount);
            }
        }

        public static List<Rotation> ParseInput(string filepath)
        {
            return 
                File.ReadAllLines(filepath)
                    .Select(line => ParseLine(line))
                    .ToList();
        }

        public static int Part1(List<Rotation> input)
        {
            var pos = 50;
            var zeroCount = 0;

            foreach(Rotation nextRotation in input)
            {
                var finalRotation = nextRotation.amount % 100;

                if (nextRotation.direction == Left)
                {
                    if (finalRotation > pos)
                    {
                        pos = 100 - (finalRotation - pos);
                    }
                    else
                    {
                        pos = pos - finalRotation;
                    }
                }
                else
                {
                    if (pos + finalRotation > 99)
                    {
                        pos = (pos + finalRotation) - 100;
                    }
                    else
                    {
                        pos = pos + finalRotation;
                    }
                }

                if(pos == 0)
                {
                    zeroCount++;
                }

            }

            return zeroCount;
        }

        public static int Part2(List<Rotation> input)
        {
            var pos = 50;
            var zeroCount = 0;

            foreach(Rotation rotation in input)
            {
                var finalAmount = rotation.amount;
                if (finalAmount > 100)
                {
                    finalAmount -= 100;
                    zeroCount++;
                }

                if (rotation.direction == Left)
                {
                    if (finalAmount > pos)
                    {
                        if (pos != 0)
                        {
                            zeroCount++;
                        }

                        pos = 100 - (finalAmount - pos);
                    }
                    else if (finalAmount == pos)
                    {
                        zeroCount++;
                        pos = 0;
                    }
                    else
                    {
                        pos -= finalAmount;
                    }
                }
                else
                {
                    if (pos + finalAmount > 99)
                    {
                        zeroCount++;
                        pos = (finalAmount + pos) - 100;
                    }
                    else
                    {
                        pos += finalAmount;
                    }
                }
            }

            return zeroCount;
        }
    }
}
